src/clients/InventoryClient.ts

```
@@ -185,15 +186,22 @@ export class InventoryClient {
   //     If this number of more than the target for the designation chain + rebalance overshoot then refund on L1.
   //     Else, the post fill amount is within the target, so refund on the destination chain.
   async determineRefundChainId(deposit: Deposit, l1Token?: string): Promise<number> {
-    const { originChainId, originToken, amount, destinationChainId } = deposit;
     const hubChainId = this.hubPoolClient.chainId;
+    const { originChainId, destinationChainId } = deposit;
 
     // Always refund on L1 if the transfer is to L1.
     if (!this.isInventoryManagementEnabled() || destinationChainId === hubChainId) {
       return destinationChainId;
     }
 
-    l1Token ??= this.hubPoolClient.getL1TokenForL2TokenAtBlock(originToken, originChainId);
+    // The InventoryClient assumes 1:1 equivalency between input and output tokens. If the tokens are not equivalent
+    // then just take repayment on the destination chain. The operator is responsible for balancing inventory.
+    const inputToken = sdkUtils.getDepositInputToken(deposit);
+    const outputToken = sdkUtils.getDepositOutputToken(deposit);
+    if (!this.hubPoolClient.areTokensEquivalent(inputToken, originChainId, outputToken, destinationChainId)) {
```

I think at this stage it might worth throwing an error here so we know if we are trying to take a refund for a fill where the tokens are not equivalent

---

Yeah OK. I was thinking that this might cause us to be open to DoS by someone making a deposit with a different outputToken (any tiny amount would suffice) but this logic is downstream of that decision, so it's probably a good call to fail hard if we make it this far.

---

Updated: https://github.com/across-protocol/relayer-v2/pull/1171/commits/828b4c0d049b16042693fc28efdf38f8c5332f94

---

Yeah, I think we should issue a warning or error log, but not cause the bot to stop. Intentional DoS or just a bad input could take the bot down.

---

I don't think we have any marginal DoS risk here because we are already filtering the array of unfilled deposits based on token equivalency, so we shouldn't actually see any swaps at this late stage. By this point the relayer has already decided it wants to make a fill and that it has sufficient balance for it, so there's no real way to say pre-empt it on that decision other than to throw.

If we were to remove or unintentionally break the token equivalency filtering then we could throw here, but then we'd have an objective error in the bot so I'm probably OK with doing that. wdyt?

---

Ah, I see, makes sense! Sorry for the noise -- I missed that in your original comment

---

All good - ++appreciate the eyes on these changes :pray: